src/EditorFeatures/CSharpTest/UseRecursivePatterns/UseRecursivePatternsTests.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UseRecursivePatterns
+{
+    public partial class UseRecursivePatternsTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+           => new CSharpUseRecursivePatternsCodeRefactoringProvider();
+
+        [Fact]
+        public async Task Test1()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.P1 == 1 [||]&& b.P2 == 2;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is
+        {
+            P1: 1, P2: 2
+        };
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test2()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return b.MetadataName == string.Empty &&
+            b.ContainingType == null &&
+            b.ContainingNamespace != null &&
+            b.ContainingNamespace.Name == string.Empty [||]&&
+            b.ContainingNamespace.ContainingNamespace == null;
+    }
+}",
+@"class C
+{
+    bool M()
+    {
+        return b is
+        {
+            MetadataName: string.Empty, ContainingType: null, ContainingNamespace:
+            {
+                Name: string.Empty, ContainingNamespace: null
+            }
+        };
+    }
+}");
+        }
+
+        [Fact]
+        public async Task Test3()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    bool M()
+    {
+        return type.Name == ""ValueTuple"" &&
+            type.ContainingSymbol is var declContainer &&
```

Comment:
semantics seem like they may have changed *slightly* (which is probably ok).  here, the `is var ...` will always succeed, and you can null on the checks ref below.  but in the converted form, it can't null ref right?

Reply:
we preserve the variable, in case it's used anywhere else (see previous comment), but semantics haven't changed because the var pattern is &&'ed with others in the source.

Reply:
1. "we preserve the variable, in case it's used anywhere else (see previous comment),"  we should probably remove the variable if not used.
2. i'm not quite getting your explanation.  previous code could have a null ref exception right?  (i.e. on declContainer.Kind).  But hte new code won't because it won't do the inner checks if ContainingSymbol is null.  Isn't that how it works?

Reply:
regarding null checks, yes it would be added by the compiler. we might also reorder expressions but property patterns are assumed to be idempotent by the compiler to allow reordering also in the generated code. this is kinda a reversed use of that assumption. there will be semantic changes if it doesn't hold.